src/main/java/uk/gov/cslearning/record/service/CourseRefreshService.java

```diff
@@ -1,92 +1,150 @@
+
+        for (String userId : userSplit.keySet()) {
+            Collection<CourseRecord> existingCourseRecords = transactionTemplate.execute(status -> courseRecordRepository.findByUserId(userId));
+            Collection<CourseRecord> userRecords = stream.replay(userSplit.get(userId), statement -> ((Activity) statement.getObject()).getId(), existingCourseRecords);
+            updatedCourseRecords.addAll(userRecords);
+        }
 
-            for (String userId : userSplit.keySet()) {
-                Collection<CourseRecord> existingCourseRecords = transactionTemplate.execute(status -> courseRecordRepository.findByUserId(userId));
-                Collection<CourseRecord> userRecords = stream.replay(userSplit.get(userId), statement -> ((Activity) statement.getObject()).getId(), existingCourseRecords);
-                updatedCourseRecords.addAll(userRecords);
+        transactionTemplate.execute(new TransactionCallbackWithoutResult() {
+            @Override
+            protected void doInTransactionWithoutResult(TransactionStatus status) {
+                courseRecordRepository.saveAll(updatedCourseRecords);
             }
+        });
+        log.info("Statements saved to DB");
+        return updatedCourseRecords.size();
+    }
+
+    private List<Statement> fetchStatementsBySinceDate(LocalDateTime since) {
+        MongoDatabase mongoDatabase = mongoClient.getDatabase(mongoProperties.getDatabase());
+        MongoCollection<Document> documents = mongoDatabase
+            .getCollection(MONGO_DOCUMENT_NAME);
+        int lastPageNumber = calculateLastPage(documents);
+        List<Statement> statements = new ArrayList<>();
 
-            transactionTemplate.execute(new TransactionCallbackWithoutResult() {
-                @Override
-                protected void doInTransactionWithoutResult(TransactionStatus status) {
-                    courseRecordRepository.saveAll(updatedCourseRecords);
-                }
-            });
-            log.info("Statements saved to DB");
-            return updatedCourseRecords.size();
-        } catch (IOException e) {
-            throw new RuntimeException("Exception retrieving xAPI statements.", e);
+        for (int pageNum = 0; pageNum < lastPageNumber; pageNum++) {
+            statements.addAll(documents.find(gte(TIMESTAMP_FIELD, Date.from(since.atZone(ZoneId.systemDefault()).toInstant())))
+                .skip(lrRefreshJobStatementBuffer * pageNum)
+                .limit(lrRefreshJobStatementBuffer)
+                .into(new ArrayList<>())
+                .stream()
+                .map(document -> mapStatement((Document) document.get(STATEMENT_FIELD)))
+                .collect(Collectors.toList()));
         }
+
+        return statements;
+    }
+
+    private int calculateLastPage(MongoCollection<Document> documents) {
+       return (int) Math.ceil((double) documents.count() / lrRefreshJobStatementBuffer);
```

Comment:
Can we test using a query in the count call (something that would return all of the documents we need, like created after 01-01-2019 or equivalent, and ideally something indexed) in case we get issues with collection metadata being slightly stale? https://docs.mongodb.com/manual/reference/method/db.collection.count/